### ISSUE TITLE
a11y: interactive controls lack accessible markup

### ISSUE THREAD

**Issue by Seirdy** [lobsters/lobsters#1091]
Many widgets cannot be reached using the keyboard (e.g. using <kbd>Tab</kbd>) and lack accessible names or ARIA states. These include:

- Upvote buttons: lack names, keyboard access, and ARIA state (already upvoted or not).
- Collapse widgets, "markdown formatting available": same as above, but also lack interactive semantics. They should have button roles.
- Comment compose box: lacks an accessible name.
- Many elements are anchors (`<a>`) despite not actually being links.

There are probably more issues, but I stopped checking at this point. There were just too many problems.

**Comment by pushcx** [comment]
Thanks for pointing it out - do you have any resources on how to label these things? It's not something we've given much attention to, but it's certainly worth it.

**Comment by Seirdy** [comment]
On Tue, May 31, 2022 at 11:59:32AM -0700, Peter Bhat Harkins wrote:
>Thanks for pointing it out - do you have any resources on how to label these things?

I'm not familiar with Rails.

I'd recommend starting with semantic HTML; use <label> for something intended to describe a form input, <button> for an interactive element that triggers an action, and <a> for an anchor that causes the agent to navigate to a link. Use <span> and <div> for inline and block-level stuff (respectively) that doesn't really have a special semantic meaning supported by HTML; this is useful for styling. Simply using the right HTML elements would be a massive improvement; the HTML alone should adequately describe the structure of the page, no styling/JS required. There's no shortage of HTML element references out there, and the WHATWG's "web developer edition" of the HTML standard has a straightforward "elements of HTML" section for a "source of truth".

Once the HTML looks semantic, a good next step would be adding some ARIA attributes: labels, states, and roles. Labels are the most important bits; semantic HTML should take care of certain states and most roles automatically, and "doing HTML's job" in ARIA is both overkill and error-prone.

MDN has a good overview of ARIA: 
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA

A reference would help; I'll ask around for existing forums that make good use of accessible markup and get back to this issue.

-- 
Seirdy
